### PR TITLE
added implementation of DRY sampler

### PR DESCRIPTION
**Pull request by l3utterfly**
As seen here: https://github.com/oobabooga/text-generation-webui/pull/5677

@p-e-w I took the liberty of porting your DRY sampler to c++ and adding it to llama.cpp

This implementation is used directly after the repeat penalty in llama.cpp, and uses the same `repeat_penalty_n` for `DRY_range`.

Thank you for the original implementation, from my tests it works amazing. It is especially useful in mobile where we can only run small quants, and smaller quants cause excessive repetition from my experience.

### CODE REVIEW

**Comment by Fristender** [comment]
Thanks for the port! Hopefully this can get merged soon.

File: llama.cpp
Comment (p-e-w) [inline llama.cpp:13049]: Unless I'm misunderstanding something, this implementation is highly inefficient.

You're looping through all candidates. Which, before truncation samplers are applied, are simply the full vocabulary of the tokenizer. For today's LLMs, that's typically between 32k and 200k.

Then in the inner loop, you're going over all tokens in the context. Which can be tens of thousands of tokens as well.

So we're talking at minimum `(candidates->size - seq_breakers_size) * last_tokens_size` iterations, which is possibly hundreds of millions. And that's not even counting the innermost sequence matching loop!

My original implementation works like this instead:

1. Find all occurrences of the last token in the context. This is typically a small percentage of the context size.
2. For each occurrence, try to extend the match backwards as far as possible. Which is usually 0-2 tokens.
3. Apply penalties to the tokens following the relevant matches.

In practice, this amounts to a single search over the context, plus a small overhead where matches are found. That's *five orders of magnitude* less work than your current implementation, assuming a vocabulary size in the tens of thousands.

**Review by p-e-w — CHANGES_REQUESTED** [review]
Thank you for getting this started! I would have made a PR here myself eventually, and in fact I might have written the original implementation for llama.cpp, but my two previous sampler-related PRs ([1](https://github.com/ggerganov/llama.cpp/pull/5561), [2](https://github.com/ggerganov/llama.cpp/pull/5675)) have received very little maintainer feedback with no way forward, and I don't enjoy putting effort into a black hole.

I have a pretty major performance concern with this implementation, which I have detailed above. Once that is addressed (which will presumably require a fairly comprehensive refactor) I will do a more in-depth review.

Regardless of the outcome of this PR, I'm happy that people like DRY! Repetition penalties are unfortunately still needed even with Llama 3, but we can do a lot better than the current default ones.

File: llama.cpp
Comment (l3utterfly) [reply llama.cpp:13049]: Thanks for the suggestions! Your method makes much more sense.

I think I didn't notice this because I was using the default top_k = 40 in llama.cpp, so it was only 40 tokens max 😂

Comment (p-e-w) [reply llama.cpp:13049]: DRY should be applied *before* any truncation samplers such as top_k. That's because you want sufficiently penalized tokens to be truncated away by those samplers. Otherwise you can be left with a distribution that violates the coherence properties that truncation is supposed to provide.

So this is another important thing that needs to be changed.

Comment (l3utterfly) [reply llama.cpp]: @p-e-w I have updated the code with your suggestions.

Looking more closely, I believe I misunderstood some of your logic in my last implementation (which was pretty naive 😂), I've refactored the sampler to better match your original implementation. Please take a look when you get the chance!

Currently, DRY is applied just after the repetition penalty, before any top k and temperature samplers.

**Comment by l3utterfly** [comment]
I messed up the merge on my branch, this one now contains a huge amount of changes. Not sure how to reset it, so I'm going to create a new PR from the latest master branch in llama.cpp.

Closing this

**Comment by l3utterfly** [comment]
Re-opened PR after force push

**Comment by p-e-w** [comment]
Perfect. Now force-push again and we should be in the clear.

File: llama.cpp
Comment (p-e-w) [inline llama.cpp:13304]: Aren't the `candidates` indices equal to the token ID? In Transformers, this is the case, which is why the original PR doesn't need to search.

If this isn't true for llama.cpp, how are the candidates ordered?

File: llama.cpp
Comment (p-e-w) [inline llama.cpp:13305]: Why isn't this in the loop initializer?

**Review by p-e-w — CHANGES_REQUESTED** [review]
I did a fairly thorough review and can confirm that, modulo some small issues, this is a faithful translation of my original Python implementation. I did not actually run the code though.

Please consider adding a separate `dry_range` parameter to control the penalty range independently of the standard repetition penalties. This is what I had originally implemented but it was removed by the maintainer; a discussion about this is ongoing in the original PR.

File: common/sampling.cpp
Comment (l3utterfly) [reply common/sampling.cpp:278]: The indentation here is trying to match the ones below from `dry_allowed_length`. What is the convention here?

image

File: llama.cpp
Comment (l3utterfly) [reply llama.cpp]: I replaced this with an `if` check instead. I'm not sure if all models use BOS tokens.

File: llama.cpp
Comment (l3utterfly) [reply llama.cpp:13311]: I looked through the creation of `candidates`. It appears it is true for this case (that the token ID = indices), but it may not always be true. It appears the candidates structure has a flag `bool sorted`, if it's true, then the `candidates` are sorted by logits descending.

We can check for that condition here? But I cannot determine if the candidates are guaranteed to have `indices = token ID` if `sorted = false`

File: common/sampling.cpp
Comment (p-e-w) [reply common/sampling.cpp]: Not sure what you mean? Both the blocks above the new parameters (ending with `penalize_nl`) and below them (starting with `prev`) have the equals signs left-aligned as closely as possible to the LHS, whereas the new parameters have three extra spaces.

But code style is really the maintainers' business. I don't care that much, just something I noticed.

File: llama.cpp
Comment (p-e-w) [reply llama.cpp:13311]: I see, I guess the purpose of sorting by logit is to simplify truncation samplers.

Probably best to keep the current code then. There are of course possible optimizations (such as interchanging the two loops and deleting tokens from `match_lengths` once they have been found, which should roughly cut the execution time in half), but I'm not sure if they are worth the extra complexity.

File: llama.cpp
Comment (p-e-w) [inline llama.cpp:13453]: The following changes are in unrelated code and probably shouldn't be in this PR.

**Review by p-e-w — APPROVED** [review]
Sampler logic seems to be correct now AFAICT. Only nits remaining, I'm too unfamiliar with the conventions of this project so I defer the rest to the maintainers.

File: common/sampling.h
Comment (l3utterfly) [reply common/sampling.h:47]: This will be converted to the maximum value of `uint32_t`, so ... uh... task failed successfully, I guess. (I will fix this)

setting `dry_penalty_last_n=-1` was to keep the same convention as `repetition_penalty`. I'll update this according to what the maintainer says.

**Comment by bakkot** [comment]
Definitely helps cut down on repetition for me. You can actually see it start to repeat a paragraph and then change course, which is fun.

Would be good to add this to [common.cpp](https://github.com/ggerganov/llama.cpp/blob/858f6b73f6e57a62523d16a955d565254be889b4/common/common.cpp) too so it's available on the CLI.

**Comment by p-e-w** [comment]
The original implementation has now been merged into text-generation-webui. There were no further algorithmic changes, so this PR implements the same sampling algorithm that is now available in TGWUI.

**Comment by belladoreai** [comment]
@l3utterfly I wrote an optimized implementation of the Python DRY sampler in oobabooga text web ui here: https://github.com/oobabooga/text-generation-webui/pull/6047

My optimizations basically eliminate the performance issues completely.

Please let me know if I can help out here in any way.

**Comment by l3utterfly** [comment]
@belladoreai thank you so much! I'll keep an eye on your original PR in oobabooga to see if the maintainer has any reviews/suggestions. If all's good, I'll port it to llama.cpp in a few days and you can do a review!

**Comment by LoopControl** [comment]
:+1: Would love to see DRY sampler get added here.

**Comment by belladoreai** [comment]
@l3utterfly DRY has now been merged into oobabooga main. Some of my optimizations got merged, some didn't get merged. You said earlier that you'd be interested in porting it. Here's some guidance:
1. Check that the data types used are appropriate (in oobabooga we were looping over values in a Torch Tensor, and it was extremely slow - at least when the tensor values were in GPU and had to be moved off the GPU one by one).
2. When you loop backwards on line 13252, set a maximum match limit to 50 (can be hardcoded, no need to parameterize). This fixes a denial-of-service vulnerability against adversarial inputs (which matters when llama.cpp is used as inference engine to server multiple users across some API, but does not matter for single user use). In oobabooga this also fixed a crash on certain kinds of repetitive inputs (related to overflow in the exponent calculation).

**Comment by ddh0** [comment]
I'd really love to see this merged, is there anything that needs to be done before that happens?

**Comment by p-e-w** [comment]
This PR has now been open for 3 full months, without a single comment from any of the maintainers, despite it being the second-most upvoted PR in this repository, and despite DRY already having been successfully integrated into text-generation-webui, SillyTavern, and KoboldCpp.

I find it disappointing that there appears to be zero interest in sampling innovations in this project, despite the massive impact they can have on output quality. Several sampling-related PRs have been open for months with little or no maintainer feedback. It looks as if this PR is going to share their fate. Meanwhile, new model architectures get merged almost on a weekly basis, with seemingly no concern for maintenance burden.

Needless to say, the maintainers are free to run this project however they want, but I'd caution @l3utterfly against investing any more time into this PR than they already have, until there is at least *some* declaration of maintainer interest, with a path towards getting this merged.

**Comment by ddh0** [comment]
> This PR has now been open for 3 full months, without a single comment from any of the maintainers, despite it being the second-most upvoted PR in this repository, and despite DRY already having been successfully integrated into text-generation-webui, SillyTavern, and KoboldCpp.
> 
> I find it disappointing that there appears to be zero interest in sampling innovations in this project, despite the massive impact they can have on output quality. Several sampling-related PRs have been open for months with little or no maintainer feedback. It looks as if this PR is going to share their fate. Meanwhile, new model architectures get merged almost on a weekly basis, with seemingly no concern for maintenance burden.
> 
> Needless to say, the maintainers are free to run this project however they want, but I'd caution @l3utterfly against investing any more time into this PR than they already have, until there is at least _some_ declaration of maintainer interest, with a path towards getting this merged.

cc for visibility: @slaren @JohannesGaessler @cebtenzzre @ikawrakow

**Comment by compilade** [comment]
> I'd really love to see this merged, is there anything that needs to be done before that happens?

@ddh0 Well, I can name a few:

- Resolve the conflicts with `master` (this is the most important)
- (if applicable) implement optimizations suggested by @belladoreai in <https://github.com/ggerganov/llama.cpp/pull/6839#issuecomment-2165312680>
- (maybe) use `int32_t` for `dry_allowed_length` and `dry_penalty_last_n`

> until there is at least some declaration of maintainer interest, with a path towards getting this merged.

I'll try to make a more in-depth review, but I'd prefer if @l3utterfly merged latest `master` onto this branch.

**Comment by ggerganov** [comment]
> I'd really love to see this merged, is there anything that needs to be done before that happens?

Also needs unit tests https://github.com/ggerganov/llama.cpp/pull/6445#issuecomment-2040373416

**Comment by wwoodsTM** [comment]
Just wanted to chime in and say that I would also love to see this merged and am gratified to those who have worked on this and for ggerganov addressing next steps. I have been regularly checking this PR for a while now and am hoping it will soon make it across the finish line. :sweat_smile: This and Min_P have been game changers as far as samplers and it is definitely I feel a critical one to add.

**Comment by ExtReMLapin** [comment]
One day we'll end up having to crowfund PRs

**Comment by l3utterfly** [comment]
> > I'd really love to see this merged, is there anything that needs to be done before that happens?
> 
> Also needs unit tests [#6445 (comment)](https://github.com/ggerganov/llama.cpp/pull/6445#issuecomment-2040373416)

I've synced with the master branch and resolved all conflicts.

Can you please give me a quick pointer on how to write unit tests for this project? Where to put them? I'm happy to write some!

**Comment by bfroemel** [comment]
> > > I'd really love to see this merged, is there anything that needs to be done before that happens?
> > 
> > 
> > Also needs unit tests [#6445 (comment)](https://github.com/ggerganov/llama.cpp/pull/6445#issuecomment-2040373416)
> 
> I've synced with the master branch and resolved all conflicts.
> 
> Can you please give me a quick pointer on how to write unit tests for this project? Where to put them? I'm happy to write some!

I think they should go in here: https://github.com/ggerganov/llama.cpp/blob/master/tests/test-sampling.cpp
And I can easily check on them, when doing a
make ./tests/test-sampling && ./tests/test-sampling

**Comment by p-e-w** [comment]
AFAICT, there is currently no logic that allows one to actually use DRY from any of the llama.cpp programs. This should probably be added, along with API support for llama-server, with an interface that supports passing sequence breakers as strings instead of token IDs (like in the original PR).

**Comment by ExtReMLapin** [comment]
Couldn't agree more, for instance, `llama-cpp-python`'s developer seems busy ATM, making it easier to implement could probably speed up "adoption" on python.

**Comment by wwoodsTM** [comment]
@l3utterfly Do you need support with any of these last parts, whether testing or what @p-e-w mentioned with CLI parameters and such to finish this PR out? My C++ skills are pretty rusty but maybe I or others can try to take a stab at supporting your efforts if you have other demands on your time atm.

**Comment by l3utterfly** [comment]
> @l3utterfly Do you need support with any of these last parts, whether testing or what @p-e-w mentioned with CLI parameters and such to finish this PR out? My C++ skills are pretty rusty but maybe I or others can try to take a stab at supporting your efforts if you have other demands on your time atm.

Thanks for your help! Any help is appreciated!

I believe you can use the dry configs with the `llama_cli` in this PR already. I will take a look at adding those parameters to the `server` in a bit

**Comment by ExtReMLapin** [comment]
Do we ? Does it need to be implemented here https://github.com/ggerganov/llama.cpp/blob/master/common/sampling.cpp#L245 ?

**Comment by l3utterfly** [comment]
> Do we ? Does it need to be implemented here https://github.com/ggerganov/llama.cpp/blob/master/common/sampling.cpp#L245 ?
> 
> `llama_sample_dry` to be referenced somewhere

The sampler is called here: https://github.com/ggerganov/llama.cpp/pull/6839/files#diff-50fcac93c9d3fb912c9f30db3657a5416d65238b22683b704cfe20071123a867R432

It goes before your "normal" temp samplers etc.

**Comment by wwoodsTM** [comment]
I've been working on it, here is what I had for server.cpp. I tried to implement it as Koboldcpp/Ooba do where either an array of strings or a JSON-encoded array of strings is accepted:

```
        slot.sparams.penalty_present    = json_value(data, "presence_penalty",   default_sparams.penalty_present);
        slot.sparams.dry_multiplier     = json_value(data, "dry_multiplier",     default_sparams.dry_multiplier);
        slot.sparams.dry_base           = json_value(data, "dry_base",           default_sparams.dry_base);
        slot.sparams.dry_allowed_length = json_value(data, "dry_allowed_length", default_sparams.dry_allowed_length);
        slot.sparams.dry_penalty_last_n = json_value(data, "dry_penalty_last_n", default_sparams.dry_penalty_last_n);
        slot.sparams.mirostat           = json_value(data, "mirostat",           default_sparams.mirostat);
        // .. skipping over some ..
        slot.sparams.min_keep           = json_value(data, "min_keep",           default_sparams.min_keep);

        // sequence breakers for DRY
        {
            auto dry_seq_breakers = data.find("dry_seq_breakers");
            if (dry_seq_breakers != data.end()) {
                try {
                    if (dry_seq_breakers->is_array()) {
                        slot.sparams.dry_seq_breakers = dry_seq_breakers->get<std::vector<std::string>>();
                    } else if (dry_seq_breakers->is_string()) {
                        slot.sparams.dry_seq_breakers = json::parse(dry_seq_breakers->get<std::string>()).get<std::vector<std::string>>();
                    } else {
                        send_error(task, "\"dry_seq_breakers\": Expected an array of strings or a JSON-encoded array of strings.", ERROR_TYPE_INVALID_REQUEST);
                        return false;
                    }
                } catch (const std::exception & e) {
                    send_error(task, std::string("\"dry_seq_breakers\": ") + e.what(), ERROR_TYPE_INVALID_REQUEST);
                    return false;
                }
            }
        }
```

**Comment by wwoodsTM** [comment]
While I have used git a good deal for personal projects, I am a complete noob with github so if I screwed something up or am not following proper etiquette with something, please accept my apologies in advance. 

So if I did this correctly, I created a PR directly on your fork with my bit of changes that I had made, both in server.cpp and in common.cpp:

[ Add DRY sampling parameters to gpt_params and server_context #29 ](https://github.com/l3utterfly/llama.cpp/pull/29)

**Comment by l3utterfly** [comment]
> While I have used git a good deal for personal projects, I am a complete noob with github so if I screwed something up or am not following proper etiquette with something, please accept my apologies in advance.
> 
> So if I did this correctly, I created a PR directly on your fork with my bit of changes that I had made, both in server.cpp and in common.cpp:
> 
> [ Add DRY sampling parameters to gpt_params and server_context #29 ](https://github.com/l3utterfly/llama.cpp/pull/29)

Thank you so much for this! It is very helpful.

I have merged your changes

**Comment by wwoodsTM** [comment]
No problem! And I realize more work was needed for the vector of strings implementation to actually work and I have been plugging away at that but hit a bug where for whatever reason the JSON settings I submit to llama-server are not coming through for the DRY settings. I have been trying to get this working before I submit again, but let me know @l3utterfly if you would like to see what I have done so far. Otherwise I will keep working on it and submit as soon as I can figure it out unless you have done any work on that already.

**Comment by wwoodsTM** [comment]
Ok, I just did a much more significant PR to your fork. I got my attempt at a somewhat optimized version with a vector of strings working:

https://github.com/l3utterfly/llama.cpp/pull/30

The one issue, as I mention on there, is that I had to move the "LLAMA_API void llama_sample_dry" function signature in llama.h to the bottom in the C++ "LLAMA_API_INTERNAL" section to avoid getting errors on compile. Obviously maintaining C compatibility in this file seems to be a priority and my use of vectors in the function signature for the API function clearly did not fit with that. My assumption is that this probably breaks certain kinds of compiles or would mess up bindings or the like? 

Maybe someone can help me fix this one issue if my implementation is felt to be worthwhile.

Finally, my use of borrowed "llama_tokenize" overloads directly placed in llama-sampling.cpp probably has a better solution for those more familiar with the overall codebase, but it was a quick and dirty way to get it working.

**Comment by l3utterfly** [comment]
@wwoodsTM Thank you! I have merged your changes

**Comment by ExtReMLapin** [comment]
From Compilade and Ggerganov's TODO list : 

- [x] Resolve the conflicts with master (this is the most important)
- [x] (if applicable) implement optimizations suggested by @belladoreai in https://github.com/ggerganov/llama.cpp/pull/6839#issuecomment-2165312680  (if i'm not wrong it's `const int MAX_CHAR_LEN = 40;`)
- [x] (maybe) use int32_t for dry_allowed_length and dry_penalty_last_n
- [x] unit tests


Good to go if everything works correctly, can't wait to test this on python-llama-cpp

**Comment by wwoodsTM** [comment]
> From Compilade and Ggerganov's TODO list :
> 
> * [x]  Resolve the conflicts with master (this is the most important)
> * [x]  (if applicable) implement optimizations suggested by @belladoreai in [added implementation of DRY sampler #6839 (comment)](https://github.com/ggerganov/llama.cpp/pull/6839#issuecomment-2165312680)  (if i'm not wrong it's `const int MAX_CHAR_LEN = 40;`)
> * [x]  (maybe) use int32_t for dry_allowed_length and dry_penalty_last_n
> * [x]  unit tests
> 
> Good to go if everything works correctly, can't wait to test this on python-llama-cpp

I would also add to that list my request of someone with more knowledge of the codebase checking the llama.h issue I mentioned above. The other things I mentioned were not as major of issues but the llama.h thing probably needs to be addressed as my hunch is that my quick “fix” there probably breaks something.

**Comment by ggerganov** [comment]
Just a heads-up, this will likely get merged after #8643

**Comment by wwoodsTM** [comment]
In looking more at this, and keeping in mind the changes coming with #8643, I thought of two possible ideas for dealing with the C interface issue I mentioned before:

**A)** We could move the sequence breakers to the `llama_sampling` object and pull them directly from there in the DRY sampling function. If I am understanding the roadmap on all this more or less correctly, this might basically go along decently with the migration of the overall sampling state to `llama_sampling` (which #8643 conveniently declares in llama.h). 

**B)** Alternatively, we could maybe preprocess the sequence breakers into an array of token arrays, refactor the DRY sampling function to only work with tokens, and maybe have the function signature in llama.h look something like this: 
```
    LLAMA_API void llama_sample_dry(
           struct llama_sampling  *  smpl,
          llama_token_data_array  *  candidates,
               const llama_token  *  last_tokens,
                          size_t     last_tokens_size,
                           float     dry_base,
                           float     dry_multiplier,
                             int     dry_allowed_length,
               const llama_token * * dry_seq_breakers,
                    const size_t  *  dry_seq_breaker_lens,
                          size_t     dry_seq_breaker_count);
```

@l3utterfly @ggerganov @p-e-w @ExtReMLapin @belladoreai Any thoughts? I would be happy to try experimenting with one of these ideas or any other suggestions in my own #8643 sandbox to try get a head start, but am hoping to get some guidance on this particular issue.

**Comment by ggerganov** [comment]
Yes, I think `llama_sampling` would be suitable to store things like sequence breakers. My immediate plan for #8643 would be to move everything remaining from current `common/sampling` into `llama_sampling`

**Comment by ddh0** [comment]
> Yes, I think `llama_sampling` would be suitable to store things like sequence breakers. My immediate plan for #8643 would be to move everything remaining from current `common/sampling` into `llama_sampling`

Now that #9294 is complete, is there a roadmap for DRY to be merged into mainline?

**Comment by wwoodsTM** [comment]
In retrospect, I think @ggerganov's previous comment about this being merged after #9294 was part vote-of-confidence and part warning, as he knew the changes that #9294 was going to introduce would likely require some significant rewriting of this PR outside of the core DRY sampling functionality. 

When I tried working on this part-way through the completion of #9294, and was looking at how extensive the changes to the overall sampling code are, I quickly realized that at least for a very rusty C++ coder like myself, it would be much wiser to just wait for the refactor to be complete before trying to do anything.

Now that it is complete, I am going to start digging into this again. I welcome any help, especially from those with more experience with the overall code-base.

**Comment by wwoodsTM** [comment]
Ok, I’m happy to say I have something really solid at this point! :sweat_smile: It’s basically complete, though I noticed some underlying issues with the sampler state not being maintained in the original refactor. My current solution "works," but it may be more of a bandaid than a true fix—definitely something to keep an eye on.

Before I proceed, I wanted to check with you, @l3utterfly: Should I submit this as a PR to this branch directly, as I did before, or would it be better to submit a brand new PR directly to [ggerganov/master](https://github.com/ggerganov/llama.cpp/tree/master) and list you as a co-author?

I assume the conflicts might be quite messy with how different the branches are now, but I know there are ways to handle that while preserving the commit history. What do you think would be the best route in this case?

**Comment by wwoodsTM** [comment]
I went ahead and made a separate PR (#9702) and listed you as co-author @l3utterfly. If you think it is better for me to undo that I can, but I hope you are ok with me being a little eager to get the ball rolling since this PR has been sitting around for a while. :smile:

**Comment by ExtReMLapin** [comment]
can be closed I guess as the other one was merged

**Pull request closed without merging**
